core/drivers/ns16550.c

```diff
@@ -70,9 +71,38 @@ static void ns16550_putc(struct serial_chip *chip, int ch)
 	serial_out(base + (UART_THR << pd->reg_shift), pd->io_width, ch);
 }
 
+static bool ns16550_have_rx_data(struct serial_chip *chip)
+{
+	struct ns16550_data *pd =
+		container_of(chip, struct ns16550_data, chip);
+	vaddr_t base = io_pa_or_va(&pd->base,
+				   (UART_LSR << pd->reg_shift) + pd->io_width);
```

Comment:
There's quite a bit of repeated code here, how about adding a `chip_to_base()` function like in the pl011.c driver and refactor a bit?

Reply:
Good suggestion, I'll try it

Reply:
It seems there are some configurable parameters included in `struct ns16550_data *pd`
We might not able to eliminate it in every function